contrib/mesos/pkg/executor/executor.go

```diff
@@ -0,0 +1,841 @@
+		client:          config.APIClient,
+		done:            make(chan struct{}),
+		outgoing:        make(chan func() (mesos.Status, error), 1024),
+		dockerClient:    config.Docker,
+		suicideTimeout:  config.SuicideTimeout,
+		kubeletFinished: config.KubeletFinished,
+		suicideWatch:    &suicideTimer{},
+		shutdownAlert:   config.ShutdownAlert,
+		exitFunc:        config.ExitFunc,
+		podStatusFunc:   config.PodStatusFunc,
+	}
+	//TODO(jdef) do something real with these events..
+	if config.Watch != nil {
+		events := config.Watch.ResultChan()
+		if events != nil {
+			go func() {
+				for e := range events {
+					// e ~= watch.Event { ADDED, *api.Event }
+					log.V(1).Info(e)
+				}
+			}()
+			k.events = events
+		}
+	}
+	return k
+}
+
+func (k *KubernetesExecutor) Init(driver bindings.ExecutorDriver) {
+	k.killKubeletContainers()
+	k.resetSuicideWatch(driver)
+	go k.sendLoop()
+	//TODO(jdef) monitor kubeletFinished and shutdown if it happens
+}
+
+func (k *KubernetesExecutor) Done() <-chan struct{} {
+	return k.done
+}
+
+func (k *KubernetesExecutor) isDone() bool {
+	select {
+	case <-k.done:
+		return true
+	default:
+		return false
+	}
+}
+
+// Registered is called when the executor is successfully registered with the slave.
+func (k *KubernetesExecutor) Registered(driver bindings.ExecutorDriver,
+	executorInfo *mesos.ExecutorInfo, frameworkInfo *mesos.FrameworkInfo, slaveInfo *mesos.SlaveInfo) {
+	if k.isDone() {
+		return
+	}
+	log.Infof("Executor %v of framework %v registered with slave %v\n",
+		executorInfo, frameworkInfo, slaveInfo)
+	if !(&k.state).transition(disconnectedState, connectedState) {
+		log.Errorf("failed to register/transition to a connected state")
+	}
+	k.initialRegistration.Do(k.onInitialRegistration)
+}
+
+// Reregistered is called when the executor is successfully re-registered with the slave.
+// This can happen when the slave fails over.
+func (k *KubernetesExecutor) Reregistered(driver bindings.ExecutorDriver, slaveInfo *mesos.SlaveInfo) {
+	if k.isDone() {
+		return
+	}
+	log.Infof("Reregistered with slave %v\n", slaveInfo)
+	if !(&k.state).transition(disconnectedState, connectedState) {
+		log.Errorf("failed to reregister/transition to a connected state")
+	}
+	k.initialRegistration.Do(k.onInitialRegistration)
+}
+
+func (k *KubernetesExecutor) onInitialRegistration() {
+	// emit an empty update to allow the mesos "source" to be marked as seen
+	k.updateChan <- kubelet.PodUpdate{
+		Pods:   []*api.Pod{},
+		Op:     kubelet.SET,
+		Source: k.sourcename,
+	}
+}
+
+// Disconnected is called when the executor is disconnected with the slave.
+func (k *KubernetesExecutor) Disconnected(driver bindings.ExecutorDriver) {
+	if k.isDone() {
+		return
+	}
+	log.Infof("Slave is disconnected\n")
+	if !(&k.state).transition(connectedState, disconnectedState) {
+		log.Errorf("failed to disconnect/transition to a disconnected state")
+	}
+}
+
+// LaunchTask is called when the executor receives a request to launch a task.
```

Comment:
Please add a few sentences to this comment walking through the steps that happened before this and that happen in this function (an indirectly through its callees). Something like "The k8sm scheduler has decided to schedule the pod (which corresponds to a Mesos Task) onto the node where this executor is running, but the binding is not recorded in the Kubernetes store yet. This function is invoked to tell the executor to record the binding in the Kubernetes store and start the pod via the Kubelet." [and if that isn't correct, please let me know and write the right thing.]